transports/webrtc/src/transport.rs

```diff
@@ -0,0 +1,697 @@
+                                err,
+                            )))));
+                        },
+                    },
+                    // Consume all events for up/down interface changes.
+                    IfWatch::Ready(watch) => {
+                        while let Poll::Ready(ev) = watch.poll_unpin(cx) {
+                            match ev {
+                                Ok(IfEvent::Up(inet)) => {
+                                    let ip = inet.addr();
+                                    if me.listen_addr.is_ipv4() == ip.is_ipv4()
+                                        || me.listen_addr.is_ipv6() == ip.is_ipv6()
+                                    {
+                                        let ma = ip_to_multiaddr(ip, me.listen_addr.port());
+                                        debug!("New listen address: {}", ma);
+                                        return Poll::Ready(Some(Ok(ListenerEvent::NewAddress(
+                                            ma,
+                                        ))));
+                                    }
+                                },
+                                Ok(IfEvent::Down(inet)) => {
+                                    let ip = inet.addr();
+                                    if me.listen_addr.is_ipv4() == ip.is_ipv4()
+                                        || me.listen_addr.is_ipv6() == ip.is_ipv6()
+                                    {
+                                        let ma = ip_to_multiaddr(ip, me.listen_addr.port());
+                                        debug!("Expired listen address: {}", ma);
+                                        return Poll::Ready(Some(Ok(
+                                            ListenerEvent::AddressExpired(ma),
+                                        )));
+                                    }
+                                },
+                                Err(err) => {
+                                    debug! {
+                                        "Failure polling interfaces: {:?}. Scheduling retry.",
+                                        err
+                                    };
+                                    me.pause = Some(Delay::new(me.sleep_on_error));
+                                    return Poll::Ready(Some(Ok(ListenerEvent::Error(
+                                        Error::IoError(err),
+                                    ))));
+                                },
+                            }
+                        }
+                    },
+                },
+                // If the listener is bound to a single interface, make sure the address reported
+                // once.
+                InAddr::One { out } => {
+                    if let Some(multiaddr) = out.take() {
+                        return Poll::Ready(Some(Ok(ListenerEvent::NewAddress(multiaddr))));
+                    }
+                },
+            }
```

Comment:
Could we extract this into its own poll implementation on `InAddr`? The conditional logic here is quite subtle (for example, the fact that we use `while let` instead of `loop { futures::ready!() }` ) and I think it would be easier to understand if we would break this down into smaller pieces.

Reply:
Agree. I copied this from tcp transport, but will be good to refactor and copy the changes back to tcp transport as well.